### PR TITLE
TRADES adversarial training protocol

### LINKED ISSUE #2031: TRADES adversarial training implementation
**Is your feature request related to a problem? Please describe.**
I want to implement an adversarial training approach TRADES because it is useful for ART in providing a better tradeoff between robustness against evasion attacks and accuracy.

**Describe the solution you'd like**
TRADES is an important adversarial training approach because it provides a better tradeoff between robustness against adversarial attacks and clean accuracy. TRADES has been proposed in paper "Theoretically Principled Trade-off between Robustness and Accuracy".

Paper link: https://proceedings.mlr.press/v97/zhang19p.html 

It is also standard component of recent and more advanced adversarial training approaches. 

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by Zaid-Hameed**
# Description

TRADES is an important adversarial training approach because it provides a better tradeoff between robustness against adversarial attacks and clean accuracy. TRADES has been proposed in paper "Theoretically Principled Trade-off between Robustness and Accuracy".

Paper link: https://proceedings.mlr.press/v97/zhang19p.html

It is also standard component of recent and more advanced adversarial training approaches.

Fixes #2031

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Trades adversarial training implementation produces results similar to original implementation
- [x] All functions in implemented code work as expected

**Test Configuration**:
- OS: Red Hat Enterprise Linux 8.7 (Ootpa)
- Python version: 3.9.12
- ART version or commit number: 1.14.1
- TensorFlow / Keras / PyTorch / MXNet version: PyTorch 1.13.1+cu117

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/defences/trainer/adversarial_trainer_trades_pytorch.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+
+import logging
+import time
+from typing import Optional, Tuple, TYPE_CHECKING
+
+import numpy as np
+from tqdm.auto import trange
+import torch
+from torch import nn
+import torch.nn.functional as F
```
Comment (beat-buesser) [inline art/defences/trainer/adversarial_trainer_trades_pytorch.py:33]: Please move the imports form torch to inside the class and methods to keep ART frame-work independent.

File: examples/adversarial_training_trades.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+opt = torch.optim.SGD(model.parameters(), lr=0.1, momentum=0.9, weight_decay=2e-4)
+lr_scheduler = MultiStepLR(opt, milestones=[74, 89], gamma=0.1)
+
+criterion = nn.CrossEntropyLoss()
+
+# Step 3: Create the ART classifier
+
+classifier = PyTorchClassifier(
+    model=model,
+    clip_values=(0.0, 1.0),
```
Comment (beat-buesser) [inline examples/adversarial_training_trades.py:132]:
```suggestion
    clip_values=(min_pixel_value, max_pixel_value),
```

File: art/defences/trainer/adversarial_trainer_trades_pytorch.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+logger = logging.getLogger(__name__)
+EPS = 1e-8
+
+
+class AdversarialTrainerTRADESPyTorch(AdversarialTrainerTRADES):
+    """
+    Class performing adversarial training following TRADES protocol.
+
+    | Paper link: https://proceedings.mlr.press/v97/zhang19p.html
+
```
Comment (beat-buesser) [inline art/defences/trainer/adversarial_trainer_trades_pytorch.py:52]:
```suggestion
```

File: art/defences/trainer/adversarial_trainer_trades_pytorch.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+    """
+
+    def __init__(self, classifier: "PyTorchClassifier", attack: "EvasionAttack", beta: float):
+        """
+        Create an :class:`.AdversarialTrainerTRADESPyTorch` instance.
+
+        :param classifier: Model to train adversarially.
+        :param attack: attack to use for data augmentation in adversarial training
+        :param beta: The scaling factor controlling tradeoff between clean loss and
+                    adversarial loss
```
Comment (beat-buesser) [inline art/defences/trainer/adversarial_trainer_trades_pytorch.py:62]:
```suggestion
        :param beta: The scaling factor controlling tradeoff between clean loss and adversarial loss.
```

File: examples/adversarial_training_trades.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    [transforms.RandomCrop(32, padding=4), transforms.RandomHorizontalFlip(), transforms.ToTensor()]
+)
+
+dataset = CIFAR10_dataset(x_train, y_train, transform=transform)
+dataloader = DataLoader(dataset, batch_size=128, shuffle=True)
+
+# Step 2: create the PyTorch model
+# model = ResNet18()
+# For running on GPU replace the model with the
+model = ResNet18().cuda()
```
Comment (beat-buesser) [inline examples/adversarial_training_trades.py:121]:
```suggestion
model = ResNet18()
```

File: art/defences/trainer/adversarial_trainer_trades_pytorch.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+
+class AdversarialTrainerTRADESPyTorch(AdversarialTrainerTRADES):
+    """
+    Class performing adversarial training following TRADES protocol.
+
+    | Paper link: https://proceedings.mlr.press/v97/zhang19p.html
+
+    """
+
+    def __init__(self, classifier: "PyTorchClassifier", attack: "EvasionAttack", beta: float):
```
Comment (beat-buesser) [inline art/defences/trainer/adversarial_trainer_trades_pytorch.py:55]: Does this work?
```suggestion
    def __init__(self, classifier: PyTorchClassifier, attack: "EvasionAttack", beta: float):
```

File: art/defences/trainer/adversarial_trainer_trades.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements adversarial training with TRADES protocol.
+
+| Paper link: https://proceedings.mlr.press/v97/zhang19p.html
+
+| It was noted that this protocol uses a modified loss. Consequently, framework specific implementations
```
Comment (beat-buesser) [inline art/defences/trainer/adversarial_trainer_trades.py:23]: Would it be possible to mention which modified loss?

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @Zaid-Hameed Thank you very much for your pull request! I have a few minor review comments, what do you think?

File: art/defences/trainer/adversarial_trainer_trades_pytorch.py (hunk 1 above)
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_trades_pytorch.py]: Done.

File: art/defences/trainer/adversarial_trainer_trades_pytorch.py (hunk 4 above)
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_trades_pytorch.py]: Done.

File: art/defences/trainer/adversarial_trainer_trades_pytorch.py [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+            )
+
+    def _batch_process(self, x_batch: np.ndarray, y_batch: np.ndarray) -> Tuple[float, float, float]:
+        """
+        Perform the operations of TRADES for a batch of data.
+        See class documentation for more information on the exact procedure.
+
+        :param x_batch: batch of x.
+        :param y_batch: batch of y.
+        """
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_trades_pytorch.py]: Added the return value description.

File: art/defences/trainer/adversarial_trainer_trades_pytorch.py (hunk 6 above)
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_trades_pytorch.py:51]: Done.

File: art/defences/trainer/adversarial_trainer_trades.py (hunk 7 above)
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_trades.py]: Included modified loss description.

**Review by beat-buesser — APPROVED** [review]
Hi @Zaid-Hameed Thank you very much and congratulations to your first PR merged to ART!

**Comment by Zaid-Hameed** [comment]
Thanks a lot @beat-buesser for all the help in it and the review. :)
